### ISSUE TITLE
releng - support arm64 docker images

### ISSUE THREAD

**Issue by nitrocode** [cloud-custodian/cloud-custodian#6102]
**Describe the bug**
Docker container fails to run policies on ARM EC2 instance type

**To Reproduce**
1. Launch an arm instance such as m6g.large
1. Run the following

    ```
    docker pull cloudcustodian/c7n:0.9.4.0
    docker run \
      --rm \
      -e AWS_DEFAULT_REGION=us-west-2 \
      -w "/opt/custodian" \
      "cloudcustodian/c7n:0.9.4.0" \
      run \
      --output-dir "output" \
      --region us-west-2 policy_test.yml
    ```
    Returns
    ```
    standard_init_linux.go:211: exec user process caused "exec format error"
    ```

**Expected behavior**

**Background (please complete the following information):**
 - OS: arm architecture on m6g.large
 - Python Version: container python
 - Custodian Version: 0.9.4
 - Cloud Provider: aws
 - Policy: N/A
 - Traceback: See above
 - `custodian version --debug` output: see container

**Additional context**
Works if I switch back to x64 architecture using m5.large

**Comment by kapilt** [comment]
Thats more a feature request then a bug, we currently publish amd64 arch images

**Comment by nitrocode** [comment]
Oh that's fantastic. How do I pull the latest arm64 c7n docker image?

**Comment by kapilt** [comment]
we currently automate docker image building with tools/dev/dockerpkg.py.. the multi arch images against the same alias are done with v2 image manifests, which is provided by buildx/buildkit tooling which is sans public api, and requires experimentally features enabled on both client and server.  buildx isn't really appropriate because we'll have builds happening independently in different envs and generally push through to serverless builds, its not clear that the extant setup with buildx is really helpful.. there's an alternative that we could potentially incorporate into dockerpkg.py, a low level v2 registry client called dxf, using dxf will allow us to patch the manifest with the appropriate image arch refs against alias in different locations.

**Comment by nitrocode** [comment]
>buildx isn't really appropriate because we'll have builds happening independently in different envs and generally push through to serverless builds

That's too bad. We've had luck building images using the following command.

```bash
docker buildx build --platform linux/arm64,linux/amd64 --push --tag custom:latest
```

Not sure I fully understand why buildx would not be appropriate. After scrolling through [dockerpkg.py](https://github.com/cloud-custodian/cloud-custodian/blob/master/tools/dev/dockerpkg.py), I can see the build process looks pretty complex.

**Comment by kapilt** [comment]
so dockerpkg is three things

- its templating/generating dockerfiles
- its building, tagging, cve scanning (via trivy), functional testing and uploading images

it does the building current via the docker client api, buildx/buildkit don't seem to have any api afaics, we could probably.. there's separate issues with buildx of ensuring its enabled on both the client and server. the implicit use of qemu by buildx does make the multi-arch building considerably more transparent wrt to host arch.  we would need to look at incorporating/ rebasing on buildx.

this was a solid deep dive on multi-arch with buildx
https://medium.com/@artur.klauser/building-multi-architecture-docker-images-with-buildx-27d80f7e2408

**Comment by kapilt** [comment]
i kicked the tires on this with buildx qemu and a graviton/arm64 instance.. we run into the issue that grpc doesn't distribute inaries for arm64 / v8 which forces a compilation which runs into the immaturity/complexity of grpc builds. i filed an upstream issue, also trying to see if upgrading setuptools inline to the docker file helps.

```
        File "/tmp/pip-req-build-uzh8pc8a/src/python/grpcio/commands.py", line 267, in build_extensions                                               [0/1487]
          support.diagnose_build_ext_error(self, error, formatted_exception)                                                                                          File "/tmp/pip-req-build-uzh8pc8a/src/python/grpcio/support.py", line 118, in diagnose_build_ext_error
          diagnostic(build_ext, error)                                                                                                                                File "/tmp/pip-req-build-uzh8pc8a/src/python/grpcio/support.py", line 97, in diagnose_attribute_error
          raise commands.CommandError(                                                                                                                              commands.CommandError: We expect a missing `_needs_stub` attribute from older versions of setuptools. Consider upgrading setuptools.
      ----------------------------------------                                                                                                                  ERROR: Command errored out with exit status 1: /usr/local/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-uzh8pc8
a/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-uzh8pc8a/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-rdw95hap/install-record.txt --single-version-exte
rnally-managed --compile --install-headers /usr/local/include/site/python3.8/grpcio Check the logs for full command output.                                                                                                                  
                                                                                                                                                                at ~/.poetry/lib/poetry/utils/env.py:948 in _run                                                                                                            
       944│                 output = subprocess.check_output(                                                                                                        945│                     cmd, stderr=subprocess.STDOUT, **kwargs                                                                                       
       946│                 )                                                  
       947│         except CalledProcessError as e:                                                                                                               →  948│             raise EnvCommandError(e, input=input_)
       949│                                                                               
       950│         return decode(output)                                                                                                                            951│                                                                    
       952│     def execute(self, bin, *args, **kwargs):                                  
                                                                                                                                                              /bin/sh: 1: cd: can't cd to tools/c7n_kube      
```

**Comment by kapilt** [comment]
grpc now supports arm.

**Comment by kapilt** [comment]
i thought this issue was handy re producing multi-arch builds within github actions 
https://github.com/actions/cache/issues/260

looking through tools/dev/dockerpkg.py it seems like we should be able to just replace the guts of build_image and push_image with subprocess calls to buildx

**Comment by thisisshi** [comment]
closed as of 0.9.19.0 https://hub.docker.com/layers/cloudcustodian/c7n/0.9.19.0/images/sha256-92629fb273575430c6cc6f18af71843a25213f3c02d8ebe4bab6f7c60515e2bd?context=explore
